Support round trip serialization of AuthenticationProperties with System.Text.Json

martincostello opened the pull request:
**PR Title**

Add attribute to support round-trip serialization of `AuthenticationProperties`

**PR Description**

Add serialization attribute to one of the constructors to support round-trip serialization of `AuthenticationProperties` with the System.Text.Json serializer.

Addresses #20722

martincostello committed:
Support roundtrip serialization with S.T.Json

Add attribute to support round-trip serialization of AuthenticationProperties with System.Text.Json.
Resolves #20722.

pranavkm commented on src/Http/Authentication.Core/test/AuthenticationPropertiesTests.cs:
```diff
@@ -302,6 +303,41 @@ public void GetBool()
+            props.Parameters.Add("baz", "quux");
+
+            var json = JsonSerializer.Serialize(props);
+            props = JsonSerializer.Deserialize<AuthenticationProperties>(json);
```
```suggestion
            var deserialized = JsonSerializer.Deserialize<AuthenticationProperties>(json);
```

pranavkm commented on src/Http/Authentication.Core/test/AuthenticationPropertiesTests.cs:
```diff
@@ -302,6 +303,41 @@ public void GetBool()
+            var json = JsonSerializer.Serialize(props);
+            props = JsonSerializer.Deserialize<AuthenticationProperties>(json);
+
+            Assert.NotNull(props);
```
```suggestion
            Assert.NotNull(deserialized);
```

pranavkm commented on src/Http/Authentication.Core/test/AuthenticationPropertiesTests.cs:
```diff
@@ -302,6 +303,41 @@ public void GetBool()
+            Assert.NotNull(props);
+
+            Assert.True(props!.AllowRefresh);
+            Assert.Equal(new DateTimeOffset(2021, 03, 28, 13, 47, 00, TimeSpan.Zero), props.ExpiresUtc);
```
```suggestion
            Assert.Equal(props.ExpiresUtc, props.ExpiresUtc);
```

pranavkm replied:
etc

martincostello committed:
Refactor test

Refactor test to not overwrite the input.

davidfowl commented on src/Http/Authentication.Abstractions/src/AuthenticationProperties.cs:
```diff
@@ -30,6 +31,7 @@ public AuthenticationProperties()
         /// Initializes a new instance of the <see cref="AuthenticationProperties"/> class.
         /// </summary>
         /// <param name="items">State values dictionary to use.</param>
+        [JsonConstructor]
```
Why this one?

martincostello replied:
I went with this one as the comment on the `Parameters` property said it wasn't intended to be used for serialisation/persistence.

I did consider adding `[JsonIgnore]` to that property, but adding the attribute here had the same effect.

pranavkm reviewed (commented):
LGTM.

@Tratcher  / @HaoK  could you sign off?

davidfowl commented:
I think this is new public API but these files we use don't support attributes...

Tratcher commented:
cc: @brockallen

HaoK commented on src/Http/Authentication.Core/test/AuthenticationPropertiesTests.cs:
```diff
@@ -302,6 +303,41 @@ public void GetBool()
+            Assert.True(deserialized.Items.ContainsKey("foo"));
+            Assert.Equal(props.Items["foo"], deserialized.Items["foo"]);
+
+            Assert.NotNull(deserialized.Parameters);
```
Maybe worth a comment saying something like // ensure parameters are not round tripped for extra clarity

martincostello committed:
Add comment for clarity

Add comment explaining that parameters are expected not to be round-tripped as suggested by code review.

Tratcher commented:
Thanks

Tratcher commented on src/Http/Authentication.Core/test/AuthenticationPropertiesTests.cs:
```diff
@@ -302,6 +303,42 @@ public void GetBool()
+
+            props.Parameters.Add("baz", "quux");
+
+            var json = JsonSerializer.Serialize(props);
```
Can you give a quick example of the serialized output? I want to make sure it doesn't contain the Parameters collection. This test only verifies the that the Parameters are not deserialized.

martincostello replied:
Sure - will post something shortly.

Tratcher replied:
Actually, we should take this further and avoid serializing any of the properties, only the Items.

martincostello replied:
#31414
